### PR TITLE
connect - core in suite-desktop

### PR DESCRIPTION
**Pull request by mroz22**
So far only some preliminary research and exploration of replacing connect-popup with suite-desktop. My general idea is to merge this feature behind debug. The motivation is that there are similar UX problems to be solved for both core-in-suite-native and core-in-suite-desktop so it makes sense to me to have some playground available for both.

suite-desktop:
- in dev or when `--expose-connect-ws` flag is on, suite-desktop opens websocket on 8090 and accepts connect calls here.
- the 1st commit has support for inter-module inducted window-focus and window-blur. it probably needs some more polishing.

connect-web:
- added new imp `core-in-suite-desktop`. what is a better name? `core-in-ws-server`?
- this impl should later be available also from base `connect` package. And also from `connect-webextension`. I need to discuss this with @karliatto 
-

image

### CODE REVIEW

File: packages/suite-desktop-core/src/app.ts [hunk 1]
```diff
@@ -195,6 +195,25 @@ const initUi = async ({
...
+        app.dock?.show();
+        if (isMacOs()) app.show();
+        if (!mainWindow.isVisible()) mainWindow.show();
+        if (mainWindow.isMinimized()) mainWindow.restore();
+        mainWindow.focus();
+    });
+
+    mainThreadEmitter.on('blur-window', () => {
+        app.hide();
+    });
```
Comment (mroz22) [inline packages/suite-desktop-core/src/app.ts]: one preliminary question - is it the correct way to achieve inter-module focus-window and blur-window? if yes, then of course code duplication needs to be solved. 

cc @marekrjpolak, @martykan

Comment (martykan) [reply packages/suite-desktop-core/src/app.ts]: We already have handlers in `window-controls` that respond to ipcMain `app/hide` and `app/focus`

Comment (martykan) [reply packages/suite-desktop-core/src/app.ts]: However it can only focus an existing window, not recreate it

File: packages/connect-web/src/impl/core-in-suite-desktop.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+     * 2. sends request to popup where the request is handled by core
+     * 3. returns response
+     */
+    public async call(params: CallMethodPayload): Promise<CallMethodAnyResponse> {
+        try {
+            await this.handshake();
+
+            this.ws?.send(
+                JSON.stringify({
+                    type: IFRAME.CALL,
```
Comment (karliatto) [inline packages/connect-web/src/impl/core-in-suite-desktop.ts:139]: Maybe using ids for messages as we use in `iframe` and new core-in-module is a good idea https://github.com/trezor/trezor-suite/blob/develop/packages/connect-web/src/impl/core-in-iframe.ts#L258

Comment (mroz22) [reply packages/connect-web/src/impl/core-in-suite-desktop.ts:139]: yes. I think we already have couple of places where we use the `send and listen for event` communication mechanism. It would be nice to find some common abstraction for it

**Comment by martykan** [comment]
Current state:
- Suite runs Websocket server at http://localhost:2335/connect-ws
- New Core mode in Connect (`suite-desktop`) that communicates with said WS
- Modal that shows info about the method being called, needs to be confirmed every time
- Security info about the process that created the connection + web origin header
Screenshot 2024-11-18 at 14 10 38

Missing features:
- Support better device modals (refactor modal state)
    - ConfirmAddressModal
    - ConfirmXpubModal
    - TransactionReviewModal
- Special methods with UI requests
	- getAccountInfo
    	- SELECT_ACCOUNT
	- composeTransaction
	    - SELECT_FEE, UPDATE_CUSTOM_FEE, INSUFFICIENT_FUNDS
- Permissions handling
  - Keep using internal Connect logic - but need different settings → Second instance of connect? Different API
  - Or skip internal Connect logic and use wrapper?

File: packages/connect-web/src/impl/core-in-suite-desktop.ts [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    Response,
+} from '@trezor/connect/src/types';
+import { ConnectFactoryDependencies, factory } from '@trezor/connect/src/factory';
+import { Login } from '@trezor/connect/src/types/api/requestLogin';
+import { createDeferred, createDeferredManager, DeferredManager } from '@trezor/utils';
+
+import { parseConnectSettings } from '../connectSettings';
+
+/**
+ * Base class for CoreInPopup methods for TrezorConnect factory.
```
Comment (mroz22) [inline packages/connect-web/src/impl/core-in-suite-desktop.ts]: comment mentioning CoreInPopup. this might be outdated

File: packages/connect-web/src/impl/core-in-suite-desktop.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                type: POPUP.HANDSHAKE,
+            }),
+        );
+        try {
+            await promise;
+        } catch (err) {
+            console.error(err);
+            throw new Error('Handshake timed out');
+        }
+    }
```
Comment (mroz22) [inline packages/connect-web/src/impl/core-in-suite-desktop.ts:70]: is this handshake actually waiting for some response?

File: packages/connect/src/constants/errors.ts [hunk 5]
```diff
@@ -78,6 +78,9 @@ export const serializeError = (payload: any) => {
     if (payload && payload.error instanceof Error) {
         return { error: payload.error.message, code: payload.error.code };
     }
+    if (payload instanceof TrezorError) {
+        return { error: payload.message, code: payload.code };
+    }
```
Comment (mroz22) [inline packages/connect/src/constants/errors.ts:83]: hmm why is this needed? 🤔

Comment (martykan) [reply packages/connect/src/constants/errors.ts:83]: It helps here
https://github.com/trezor/trezor-suite/pull/14690/files/b1630f7dfba7a66977e205e605b1c41372282c97#diff-2e9d057f0bfe2cc92fe50d4ce28838622d9e79fcca010ab8847a0fa288da13fdR245
For some reason the original method serializes payload.error, not the actual error

File: packages/connect-web/src/impl/core-in-suite-desktop.ts (hunk 4 above)
Comment (martykan) [reply packages/connect-web/src/impl/core-in-suite-desktop.ts:70]: Yeah, the response is received by `this.ws.addEventListener('message', ...)` like all other messages

**Review by mroz22 — COMMENTED** [review]
# 🦭 ✅

**Review by martykan — APPROVED** [review]
